### DISCUSSION TITLE
ORT include concept for monorepo

### DISCUSSION THREAD

**Discussion by nnobelis** [oss-review-toolkit/ort#9290] [Ideas]
We have a gigantic monorepo containing several modules.

Our users want to scan only one module corresponding to the software they are delivering.

ORT excludes are not scaling too well for such use cases as you have to write a lot of them. Additionally, when a new directory is added to the monorepo, it is taken by default.

Has the ORT team ever considered the concept of ORT includes (allow list) i.e. everything is ignored by default and you have to specify what do you want to include in your run, for instance with some ORT includes ?

**Comment by sschuberth** [comment]
Have you alternatively considered to [limited the initial cloning of the project](https://www.baeldung.com/ops/git-clone-subdirectory) to scan to certain sub-directories via partial clones / sparse checkouts?

**Reply by nnobelis** [reply]
So you mean skip ORT downloader and manually prepare the repository to analyze with `git sparse-checkout` ?

**Reply by sschuberth** [reply]
There should be no need to skip / not use the ORT downloader. The Git VCS plugin [already supports sparse checkouts](https://github.com/oss-review-toolkit/ort/blob/8c103c4d4dc6a829a4d7d6d67a9b44147686b500/plugins/version-control-systems/git/src/main/kotlin/Git.kt#L148-L158), and [the downloader command already has `--vcs-path`](https://github.com/oss-review-toolkit/ort/blob/8c103c4d4dc6a829a4d7d6d67a9b44147686b500/plugins/commands/downloader/src/main/kotlin/DownloaderCommand.kt#L141-L145) which should trigger that code path. So I guess you could just use that as-is.

**Reply by nnobelis** [reply]
Unfortunately you would only "cover" one directory of the monorepo with the `vcs-path`.

I am thinking of more complex use cases where you need to include various source directory (and excludes the others).

**Reply by sschuberth** [reply]
Paths could be globs, and also at least Git underneath supports multiple paths. You'd just need to ensure to supported that in the CLI as well.

**Reply by nnobelis** [reply]
To be sure I understand you right, you mean put a glob in the `vcs-path` and use the downloader ?

**Reply by sschuberth** [reply]
Yes.

**Reply by nnobelis** [reply]
I will have a look but I am not convinced. I have the guts feeling that a lot of different directories must be selected, so putting everything in a glob would not scale well.

**Reply by sschuberth** [reply]
You could also make `--vcs-path` take a comma-separated list of globs, if that helps.
